pkg/ingester/ingester.go

```diff
@@ -2492,6 +2492,60 @@ func (i *Ingester) compactBlocks(ctx context.Context, force bool, allowed *util.
 	})
 }
 
+// compactBlocksToReduceInMemorySeries checks whether the per-tenant TSDB Head should be compacted into a block
+// in order to reduce the in-memory series.
+func (i *Ingester) compactBlocksToReduceInMemorySeries(ctx context.Context, now time.Time) {
+	// Skip if disabled.
+	if i.cfg.BlocksStorageConfig.TSDB.EarlyHeadCompactionMinInMemorySeries <= 0 || !i.cfg.ActiveSeriesMetrics.Enabled {
+		return
+	}
+
+	// No need to prematurely compact TSDB heads if the number of in-memory series is below a critical threshold.
+	numInitialMemorySeries := i.seriesCount.Load()
+	if numInitialMemorySeries < i.cfg.BlocksStorageConfig.TSDB.EarlyHeadCompactionMinInMemorySeries {
+		return
+	}
+
+	level.Info(i.logger).Log("msg", "the number of in-memory series is higher than the configured early compaction threshold", "in_memory_series", numInitialMemorySeries, "early_compaction_threshold", i.cfg.BlocksStorageConfig.TSDB.EarlyHeadCompactionMinInMemorySeries)
+
+	var (
+		usersToCompact          []string
+		forcedCompactionMaxTime = now.Add(-i.cfg.ActiveSeriesMetrics.IdleTimeout).UnixMilli()
+	)
+
+	// Find a list of tenants that could help to significantly drop the number of in-memory series
+	// if we would forcely compact their TSDB head.
+	for _, userID := range i.getTSDBUsers() {
+		db := i.getTSDB(userID)
+		if db == nil {
+			continue
+		}
+
+		// Purge the active series so that the next call to Active() will return the up-to-date count.
+		db.activeSeries.Purge(now)
+
+		// Estimate the number of series that would be dropped from the TSDB Head if we would
+		// compact the head up until "now - active series idle timeout".
+		estimatedSeriesReduction := db.Head().NumSeries() - uint64(db.activeSeries.Active())
+
+		// Skip if the estimated series reduction is too low (there would be no big benefit).
+		if float64(estimatedSeriesReduction)/float64(numInitialMemorySeries)*100 < float64(i.cfg.BlocksStorageConfig.TSDB.EarlyHeadCompactionMinEstimatedSeriesReductionPercentage) {
```

Comment:
Shouldn't we do this?

```suggestion
		if float64(estimatedSeriesReduction)/float64(db.Head().NumSeries())*100 < float64(i.cfg.BlocksStorageConfig.TSDB.EarlyHeadCompactionMinEstimatedSeriesReductionPercentage) {
```

Reply:
The idea of this PR is to check it against the total number of series in the ingester, and not the per-tenant TSDB. Let's say you configure the percentage to 10%, a tenant has 100 series, and 20 are inactive. If you check against the per-tenant TSDB, it will trigger the compaction, but in practice it will not move a needle. Comparing against the total number of series in the ingester allow us to only trigger the early compaction in case there are big opportunities.

Reply:
But then this setting will be quite useless in multi-tenant environments. If you have 10 tenants each one saving 50% of their series, you could save a total of 50% by compacting them, but with current logic you won't trigger the compaction because each tenant will save only 5% of the total. Right?

Reply:
The math is correct but I think it's all about the percentage you configure. Maybe in multi-tenant environments it makes sense a lower %. I think this is something we can refine over time, but I definitely want to keep a way to avoid compacting a TSDB when the overall impact is negligible.

Reply:
I think that having it this way will require tons of tuning, because the setting doesn't relate to the final effect. I have an ingester with a number of in-memory series, and I want to make sure that we compact heads when I can save a given amount of memory. I shouldn't need to update the percentage based on the number of tenants I have in order to achieve that. 

IMO it's counter-intuitive.

OTOH, what's the problem about compacting 10 smaller Heads vs compacting one bigger?

Reply:
Discussed offline. I pushed a [commit](https://github.com/grafana/mimir/pull/5371/commits/637d14ea98142c989bc74abd810246aef86dc41c) to include Oleg's suggestion, but slightly modified to also compact all TSDBs with an high number of inactive series (even if below the configured %).

There's no strong consensus between the two of us on what's the right way to decide which TSDBs to compact. The gist is that is difficult to have good arguments because we'll learn it the hard way (running it for some time).